app/_src/kubernetes-ingress-controller/guides/security/verify-upstream-tls.md

```diff
@@ -0,0 +1,417 @@
+              {
+                "containerPort": 443
+              }
+            ],
+            "env": [
+              {
+                "name": "HTTPS_PORT",
+                "value": "443"
+              },
+              {
+                "name": "TLS_CERT_FILE",
+                "value": "/etc/tls/tls.crt"
+              },
+              {
+                "name": "TLS_KEY_FILE",
+                "value": "/etc/tls/tls.key"
+              }
+            ],
+            "volumeMounts": [
+              {
+                "mountPath": "/etc/tls",
+                "name": "tls"
+              }
+            ]
+          }
+        ],
+        "volumes": [
+          {
+            "name": "tls",
+            "secret": {
+              "secretName": "goecho-tls"
+            }
+          }
+        ]
+      }
+    }
+  }
+}'
+```
+
+Also, patch the service to use HTTPS by adding the `konghq.com/protocol: https` annotation and the `spec.ports`
+entry.
+
+```shell
+kubectl patch service echo -p '{
+  "metadata": {
+    "annotations": {
+      "konghq.com/protocol": "https"
+    }
+  },
+  "spec": {
+    "ports": [
+      {
+        "name": "https",
+        "port": 443,
+        "targetPort": 443
+      }
+    ]
+  }
+}'
+```
+
+The results should look like this:
+
+```text
+deployment.apps/echo patched
+service/echo patched
+```
+
+## Expose the goecho service
+
+Now that the `kong/goecho` service is serving HTTPS, we need to expose it.
+
+{% include /md/kic/http-test-routing-resource.md service=echo port=443 route_type=PathPrefix %}
+
+Verify connectivity by issuing an HTTP request to proxy. The service serves HTTPS but {{ site.base_gateway }} initiates
+the connection and proxies it as HTTP in this case, thus the request should be made over HTTP. The `Host` header is
+required to match the hostname of the service.
+
+```shell
+curl -H Host:kong.example $PROXY_IP/echo
+```
+
+You should see a response similar to this:
+
+```text
+Welcome, you are connected to node orbstack.
+Running on Pod echo-bd94b7dcc-qxs2b.
+In namespace default.
+With IP address 192.168.194.9.
+Through HTTPS connection.
+```
+
+That means the service is up and running and {{ site.base_gateway }} connects to it successfully over HTTPS, _without
+verification_.
+
+## Configure {{ site.base_gateway }} to verify the upstream TLS certificate
+
+### Enable TLS verification
+
+To configure {{ site.base_gateway }} to verify the certificate of the upstream service, we need to annotate the
+service accordingly.
+
+```shell
+kubectl annotate service echo konghq.com/tls-verify=true
+```
+
+The results should look like this.
+
+```text
+service/echo annotated
+```
+
+Now, when you issue the same request as before, you should see an error similar to this.
+
+```shell
+curl -H Host:kong.example $PROXY_IP/echo
+```
+
+```text
+{
+  "message":"An invalid response was received from the upstream server",
+  "request_id":"e2b3182856c96c23d61e880d0a28012f"
+}
+```
```

Comment:
So I was confused because this didn't workout.

I had to either wait a long time or restart the gateway for the change to work. My guess is that Kong keeps the upstream connection around.
I think it's more likely this than anything because switch it back to false is almost instant.

Here are the logs for:

true --> false --> true:

```
2024/12/13 10:12:10 [error] 1337#0: *1454 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "6336bb40e18739365ba3b32de7b8e512"
2024/12/13 10:12:10 [error] 1337#0: *1454 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "6336bb40e18739365ba3b32de7b8e512"
2024/12/13 10:12:10 [error] 1337#0: *1454 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "6336bb40e18739365ba3b32de7b8e512"
2024/12/13 10:12:10 [error] 1337#0: *1454 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "6336bb40e18739365ba3b32de7b8e512"
2024/12/13 10:12:10 [error] 1337#0: *1454 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "6336bb40e18739365ba3b32de7b8e512"
2024/12/13 10:12:10 [error] 1337#0: *1454 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "6336bb40e18739365ba3b32de7b8e512"
10.244.0.1 - - [13/Dec/2024:10:12:10 +0000] "GET /echo HTTP/1.1" 502 126 "-" "curl/8.7.1" kong_request_id: "6336bb40e18739365ba3b32de7b8e512"
10.244.0.1 - - [13/Dec/2024:10:12:11 +0000] "GET /echo HTTP/1.1" 200 169 "-" "curl/8.7.1" kong_request_id: "3f5f2614508d8933e246f52fbd10a45e"
10.244.0.1 - - [13/Dec/2024:10:12:12 +0000] "GET /echo HTTP/1.1" 200 169 "-" "curl/8.7.1" kong_request_id: "2b60920255b82b68a786e9779edae94d"
10.244.0.1 - - [13/Dec/2024:10:12:13 +0000] "GET /echo HTTP/1.1" 200 169 "-" "curl/8.7.1" kong_request_id: "e98ff21e43794d283b6a5a6a9376e56d"
10.244.0.1 - - [13/Dec/2024:10:12:21 +0000] "GET /echo HTTP/1.1" 200 169 "-" "curl/8.7.1" kong_request_id: "d6b8b3060a94b5306b50362a64eab701"
10.244.0.1 - - [13/Dec/2024:10:12:22 +0000] "GET /echo HTTP/1.1" 200 169 "-" "curl/8.7.1" kong_request_id: "b2ea3e688733ca1bac33b11cda41c826"
10.244.0.1 - - [13/Dec/2024:10:12:22 +0000] "GET /echo HTTP/1.1" 200 169 "-" "curl/8.7.1" kong_request_id: "ecc67c1586b38e5c4cd4af747cd9e04c"
10.244.0.1 - - [13/Dec/2024:10:12:41 +0000] "GET /echo HTTP/1.1" 200 169 "-" "curl/8.7.1" kong_request_id: "a52b03712d35ab7d01caeb480862be31"
2024/12/13 10:13:54 [error] 1337#0: *1505 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "0ebde4335d6d9f76ef1aa4444d1c0b48"
2024/12/13 10:13:54 [error] 1337#0: *1505 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "0ebde4335d6d9f76ef1aa4444d1c0b48"
2024/12/13 10:13:54 [error] 1337#0: *1505 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "0ebde4335d6d9f76ef1aa4444d1c0b48"
2024/12/13 10:13:54 [error] 1337#0: *1505 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "0ebde4335d6d9f76ef1aa4444d1c0b48"
2024/12/13 10:13:54 [error] 1337#0: *1505 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "0ebde4335d6d9f76ef1aa4444d1c0b48"
2024/12/13 10:13:54 [error] 1337#0: *1505 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "0ebde4335d6d9f76ef1aa4444d1c0b48"
10.244.0.1 - - [13/Dec/2024:10:13:54 +0000] "GET /echo HTTP/1.1" 502 126 "-" "curl/8.7.1" kong_request_id: "0ebde4335d6d9f76ef1aa4444d1c0b48"
2024/12/13 10:13:57 [error] 1337#0: *1512 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "adfa282ad26199740b2daf26d85abb2c"
2024/12/13 10:13:57 [error] 1337#0: *1512 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "adfa282ad26199740b2daf26d85abb2c"
2024/12/13 10:13:57 [error] 1337#0: *1512 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "adfa282ad26199740b2daf26d85abb2c"
2024/12/13 10:13:57 [error] 1337#0: *1512 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "adfa282ad26199740b2daf26d85abb2c"
2024/12/13 10:13:57 [error] 1337#0: *1512 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "adfa282ad26199740b2daf26d85abb2c"
2024/12/13 10:13:57 [error] 1337#0: *1512 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 10.244.0.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://10.244.0.8:443/", host: "kong.example", request_id: "adfa282ad26199740b2daf26d85abb2c"
10.244.0.1 - - [13/Dec/2024:10:13:57 +0000] "GET /echo HTTP/1.1" 502 126 "-" "curl/8.7.1" kong_request_id: "adfa282ad26199740b2daf26d85abb2c"
```

The change back to true was done at 12:15. It might be worth just pointing out that it needs to time with no traffic.